app/components/annotation/Article.js

```diff
@@ -17,14 +59,45 @@ export default React.createClass({
     let article = this.props.article;
     let topic = this.props.topics[topic_id];
 
+    var text = this.props.article.text;
+    var highlights = this.props.highlights || [];
+
+    var start = 0;
+    var tail = '';
+    var l = highlights.length;
+
+    if (l === 0) {
+      tail = text;
+    } else if (highlights[l - 1].end !== text.length) {
+      tail = <span>{text.substring(highlights[l - 1].end, text.length)}</span>;
+    }
+
     return (
       <div>
         <div className='tua__header-text'>
           Focus on the bold text about '{topic.name}' and answer the questions.
         </div>
-        <div className='article'>{article.text}</div>
+        <div ref={(ref) => this.articleRef = ref} className='article' onClick={this.handleClick}>
+          {Array(highlights.length * 2).fill().map((_,i) => {
```

Comment:
I need to iterate `highlights.length * 2` times here, but react disallows control flow within JSX (for good reason), so this is my way of creating a for loop. `_` is my goto throwaway iteration variable name, and it's not being used here. i agree that this isn't clear on first read but it's the cleanest of the hacks to get the right iteration

... the reason I need to iterate `n*2` times is because I need to return two `<span>`s inside here, but if return them both in the same statement i have to wrap them because of JSX rules, which makes iterating through them to count text offsets much more difficult.